pkg/ebpf/bytecode/runtime/log_debug_test.go

```diff
@@ -0,0 +1,95 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package runtime
+
+import (
+	"bufio"
+	"math"
+	"os"
+	"testing"
+
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/ebpftest"
+	manager "github.com/DataDog/ebpf-manager"
+	"github.com/DataDog/ebpf-manager/tracefs"
+	"github.com/cilium/ebpf"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sys/unix"
+)
+
+//go:generate $GOPATH/bin/include_headers pkg/ebpf/testdata/c/runtime/logdebug-test.c pkg/ebpf/bytecode/build/runtime/logdebug-test.c pkg/ebpf/c pkg/network/ebpf/c/runtime pkg/network/ebpf/c
+//go:generate $GOPATH/bin/integrity pkg/ebpf/bytecode/build/runtime/logdebug-test.c pkg/ebpf/bytecode/runtime/logdebug-test.go runtime
+
+func TestPatchPrintkNewline(t *testing.T) {
+	ebpftest.TestBuildMode(t, ebpftest.RuntimeCompiled, "", func(t *testing.T) {
+		cfg := ddebpf.NewConfig()
+		require.NotNil(t, cfg)
+
+		buf, err := LogdebugTest.Compile(cfg, []string{"-g", "-DDEBUG=1"}, nil)
+		require.NoError(t, err)
+		defer buf.Close()
+
+		idPair := manager.ProbeIdentificationPair{
+			EBPFFuncName: "logdebugtest",
+			UID:          "logdebugtest",
+		}
+		mgr := &manager.Manager{
+			Probes: []*manager.Probe{
+				{
+					ProbeIdentificationPair: idPair,
+					XDPAttachMode:           manager.XdpAttachModeSkb,
+					IfIndex:                 1,
+				},
+			},
+		}
+
+		opts := manager.Options{
+			RLimit: &unix.Rlimit{
+				Cur: math.MaxUint64,
+				Max: math.MaxUint64,
+			},
+			MapEditors: make(map[string]*ebpf.Map),
+			ConstantEditors: []manager.ConstantEditor{
+				ddebpf.GetPatchedPrintkEditor(),
+			},
+		}
+
+		tp, err := tracefs.OpenFile("trace_pipe", os.O_RDONLY, 0)
+		require.NoError(t, err)
+		traceReader := bufio.NewReader(tp)
+		t.Cleanup(func() { _ = tp.Close() })
+
+		require.NoError(t, mgr.InitWithOptions(buf, opts))
+		require.NoError(t, mgr.Start())
+		t.Cleanup(func() { mgr.Stop(manager.CleanAll) })
+
+		progs, ok, err := mgr.GetProgram(idPair)
+		require.NoError(t, err)
+		require.True(t, ok)
+		require.NotNil(t, progs)
+		require.NotEmpty(t, progs)
+		prog := progs[0]
+
+		input := make([]byte, 30) // Enough for the arguments
+		ret, _, err := prog.Test(input)
```

Comment:
The documentation for `Test` states that it requires at least kernel `4.12`. CI tests kernels all the way down to `4.4`. Should we skip this test on those kernels?

Reply:
Yes, I pushed this test to see how it behaved but I think it's way too unreliable to have it in CI. Some systems don't have the trace pipe enabled by default, in others it seems to hang after a certain amount of output... In this cases do you leave the test with a t.Skip at the start or just delete it altogether?

Reply:
From the ci runs below, it does look like it is pretty unreliable. In this case I would not merge this, and maybe work on making this more reliable at a later time.

Reply:
Rather than `prog.Test`, we could use a kprobe on `do_vfs_ioctl` and trigger it from the test using the `ioctl` syscall and a custom number. See `pkg/security/probe/erpc/erpc.go` for an example of how to do this, but we won't need any of the `Request` data structure.

This should allow us to support more kernels.